### PR TITLE
WIP: resolve property references to other namespace

### LINKED ISSUE #36: Exception in DescriptorBuilder when inherited properties have the same name
Hi,

the `DescriptorBuilder` throws an exception if a property with the same name is defined in the inheritance chain of more than one of its super classes:

```
Error: property <any> already defined; override of <mh:AnotherRoot#mh:any> by <props:Root#props:any> not allowed without redefines
      at DescriptorBuilder._fd7‍.r.DescriptorBuilder.assertNotDefined (lib/descriptor-builder.js:180:11)
      at DescriptorBuilder._fd7‍.r.DescriptorBuilder.addNamedProperty (lib/descriptor-builder.js:138:10)
      at DescriptorBuilder._fd7‍.r.DescriptorBuilder.addProperty (lib/descriptor-builder.js:53:8)
      at DescriptorBuilder.<anonymous> (lib/descriptor-builder.js:227:12)
      at forEach (node_modules/min-dash/dist/index.js:170:20)
      at DescriptorBuilder._fd7‍.r.DescriptorBuilder.addTrait (lib/descriptor-builder.js:202:3)
      at lib/registry.js:185:13
      at Registry._0ac‍.r.Registry.mapTypes (lib/registry.js:165:3)
      at traverseSuper (lib/registry.js:155:10)
      at forEach (node_modules/min-dash/dist/index.js:170:20)
      at Registry._0ac‍.r.Registry.mapTypes (lib/registry.js:162:3)
      at Registry._0ac‍.r.Registry.getEffectiveDescriptor (lib/registry.js:184:8)
      at Moddle._1b8‍.r.Moddle.getType (lib/moddle.js:97:32)
      at Moddle._1b8‍.r.Moddle.create (lib/moddle.js:65:19)
      at Context.<anonymous> (test/spec/properties.js:589:34)
      at processImmediate (internal/timers.js:461:21)
```

I [created a test case](https://github.com/fra-pa/moddle/commit/b1b9795512d67f3bab3e7135068ad52f8c552f1d) with a simplified hierarchy:
multiple-inherited-properties

`mh:MultipleInherited` has two properties named `any`:
- `mh:AnotherRoot#any`
- `props:Root#any`

and the `DescriptorBuilder` does not map them into seperate namespaces.

In my use-case, the problem occurs with the [UMLDI](https://www.omg.org/spec/UMLDI/) meta-model, where all the diagram types inherit from `UMLDI:UMLDiagram`:
umldi-inheritance

Regards,
Frank

### PR DESCRIPTION
**Pull request by fra-pa**
This fixes the [issue](#36) with equally named properties defined in inherited types as long as these types have different namespaces.
DescriptorBuilder to lookup properties in other namespaces if the name is explicitly prefixed (i.e. namespace:propertyName).

Naming conflicts resulting from multiple inheritance within the same model are still NOT resolved. E.g.:

multiple-inheritance-same-model

If the scope for namespaces would be expanded to types, the properties could be seperated and mapped accordingly. But I doubt that this is necessary (or even useful) as long as *moddle* has no explicit namespacing (e.g. with a new keyword; which could be a useful extension). IMO, a proper model design should avoid this kind of conflicts and it seems to be quite some effort to implement.

Closes #36.

### CODE REVIEW

File: test/spec/properties.js [hunk 1]
```diff
@@ -455,4 +467,140 @@ describe('properties', function() {
...
+          // when
+          var instance = mhModel.create('mh:MultipleInherited');
+          instance.set('mh:any', [ 'test' ]);
+          var originalProperty = instance.get('any');
+          var localProperty = instance.get('mh:any');
+          var otherProperty = instance.get('props:any');
+
+          // then
+          expect(originalProperty.length).to.equal(1);
+          expect(localProperty.length).to.equal(1);
```
Comment (nikku) [inline test/spec/properties.js:580]: Use actual strict equality test here, i.e. `expect(originalProperty).to.equal(expectedProperty)`.

File: test/spec/properties.js [hunk 2]
```diff
@@ -455,4 +467,140 @@ describe('properties', function() {
...
+        it('access via original name', function() {
+
+          // given
+          var instance = mhModel.create('mh:MultipleInherited');
+
+          // when
+          var property = instance.get('any');
+
+          // then
+          expect(property).to.exist;
```
Comment (nikku) [inline test/spec/properties.js:518]: Verify actual property value, i.e. `expect(property).to.eql([])`.

TODO: Verify actual property value on non-many fields, too. Include test on default values, too.

File: test/spec/properties.js [hunk 3]
```diff
@@ -73,6 +73,18 @@ describe('properties', function() {
...
+
+    it('should NOT add already defined property without redefine', function() {
+
+      // when
+      var getType = function() {
+        model.getType('props:BaseWithAlreadyDefinedId');
+      };
+
+      // then
+      expect(getType).to.throw(Error);
```
Comment (nikku) [inline test/spec/properties.js:85]: TODO: Verify which error is thrown `expect(getType).to.throw(/some erorr message matcher/)`.

Why do we blow up late, i.e. on instantiation only? I'd assume we'd blow up when creating the package.

File: lib/properties.js [hunk 4]
```diff
@@ -61,11 +66,16 @@ Properties.prototype.get = function(target, name) {
     return target.$attrs[name];
   }
 
-  var propertyName = property.name;
+  var propertyName = getPropertyName(target, name, property);
 
   // check if access to collection property and lazily initialize it
   if (!target[propertyName] && property.isMany) {
-    defineProperty(target, property, []);
+    Object.defineProperty(target, propertyName, {
```
Comment (nikku) [inline lib/properties.js:73]: What is the reason to not use the existing utility?

File: test/spec/properties.js [hunk 5]
```diff
@@ -455,4 +467,140 @@ describe('properties', function() {
...
+
+        // when
+        var Type = mhModel.getType('mh:MultipleInherited');
+        var descriptor = model.getElementDescriptor(Type);
+
+        // then
+        expect(Type).to.exist;
+        expect(descriptor).to.exist;
+        expect(descriptor.propertiesByName.any).to.exist;
+        expect(descriptor.propertiesByName['mh:any']).to.exist;
```
Comment (nikku) [inline test/spec/properties.js:490]: TODO: Verify which descriptor is equal (strict equality) to which descriptor.

`expect(descriptor.propertiesByName.any).to.equal(descriptor.propertiesByName['mh:any'])`

Comment (nikku) [reply test/spec/properties.js:490]: TODO: Add test to verify the following:

* Package a defines `a:Foo#a`
* Package b defines `b:Other#a`
* Package c defines `c:MultipleInherited { a:Foo, b:Other }`
* :arrow_right: What is the local descriptor for `c.MultipleInherited#a`?

File: lib/descriptor-builder.js [hunk 6]
```diff
@@ -172,15 +174,18 @@ DescriptorBuilder.prototype.setIdProperty = function(p, validate) {
...
 
-DescriptorBuilder.prototype.assertNotDefined = function(p, name) {
-  var propertyName = p.name,
+DescriptorBuilder.prototype.assertNotDefined = function(p) {
+  var ns = p.ns,
+      propertyName = p.name,
       definedProperty = this.propertiesByName[propertyName];
 
-  if (definedProperty) {
+  if (this.propertiesByName[ns.name] && this.propertiesByName[ns.localName]) {
```
Comment (nikku) [inline lib/descriptor-builder.js:182]: What should this check accomplish? Is there a situation where the property has been defined within the same `ns` and no local name assigned? If so, we should verify that would be actually fine.

**Comment by nikku** [comment]
Thanks for opening this PR @fra-pa.

I was able to have a look and added some comments how to improve the test coverage. The overall code changes look simple enough at the moment. We'd need to improve the test coverage though and test [some more edge cases](https://github.com/bpmn-io/moddle/pull/39#pullrequestreview-660722122).

**Comment by fra-pa** [comment]
Thanks for review and feedback @nikku. I'll improve testing as soon as I find time.

One question regarding your process with commits to the PR branch. Should I just add them or squash into a single one (with forced push or a new branch)?

**Comment by nikku** [comment]
You can just add additional commits, as you amend your PR. Using some `git` magic you or myself can always do the cleanup later. Adding commits on top makes it easier for me as a reviewer to follow along the progress of your contribution.

File: test/spec/properties.js (hunk 1 above)
Comment (fra-pa) [reply test/spec/properties.js]: > Use actual strict equality test here, i.e. expect(originalProperty).to.equal(expectedProperty).

Changed this and also use `eql()` for deep-testing 'many's now.

File: test/spec/properties.js (hunk 2 above)
Comment (fra-pa) [reply test/spec/properties.js]: > Verify actual property value, i.e. expect(property).to.eql([]).

Good point. I keep one test for existance to fail early in case property is undefined.

> TODO: Verify actual property value on non-many fields, too.

Added,

Comment (fra-pa) [reply test/spec/properties.js]: > Include test on default values, too.

You nailed it with this. [Added](https://github.com/bpmn-io/moddle/pull/39/commits/c332a7eaaecd0ddefa6b184120f6ff9fc2069110) the test and it fails!

Fixed it [here](https://github.com/bpmn-io/moddle/pull/39/commits/34457263442fd98c38df1e2e954fa8425a0b4492).
Sorry for the CI failure. I overlooked the missing semicolon because I also see linting errors for `.to.exist`, as this is not a function. Maybe add *dirty-chai* module for *eslint*? Also [fixed](https://github.com/bpmn-io/moddle/pull/39/commits/1b5ce8c45816ab50c8d112970a7ca932b04fc980) this of course.

File: lib/factory.js [hunk 7]
```diff
@@ -27,7 +27,17 @@ Factory.prototype.createType = function(descriptor) {
   // initialize default values
   forEach(descriptor.properties, function(p) {
     if (!p.isMany && p.default !== undefined) {
-      prototype[p.name] = p.default;
+      var name;
+
+      if (p.ns.prefix === descriptor.ns.prefix) {
+
+        // use local name if property already in local ns
+        name = p.name;
```
Comment (nikku) [inline lib/factory.js:35]: Interesting question here: What I'd expect is to have local name _and_ ns name attribute present.

I.e. I can always access the value via `foo["a:id"]` regardless if it has a local name set or not.

Likely this is only possible to accomplish via proxies, am I right about that?

**Comment by nikku** [comment]
Overall I believe we make good progress on this PR. 

What I'd love to see is whether our existing [serialization utility](https://github.com/bpmn-io/moddle-xml) is actually able to consume multiple-inherited diagrams, modify them _and_ properly serialize them. Only then we're talking about an actual meaningful roundtrip that would justify adding the complexity to this library.

File: test/spec/properties.js (hunk 3 above)
Comment (fra-pa) [reply test/spec/properties.js]: > TODO: Verify which error is thrown expect(getType).to.throw(/some erorr message matcher/).

[Changed](https://github.com/bpmn-io/moddle/pull/39/commits/7bdb05fa5f70fc9df6e15d485378b63af85c3c61). Compares the whole exception string now.

> Why do we blow up late, i.e. on instantiation only? I'd assume we'd blow up when creating the package.

Yep, I'd also prefere earlier exception. If not `new Moddle()`, then latest in `Moddle.create()`.
This needs some kind of verification though. Maybe we should address this in a seperate issue?

This also applies to #38.

File: test/spec/properties.js (hunk 5 above)
Comment (fra-pa) [reply test/spec/properties.js]: > TODO: Verify which descriptor is equal (strict equality) to which descriptor.
>
> `expect(descriptor.propertiesByName.any).to.equal(descriptor.propertiesByName['mh:any'])``

Another hit! [Added the test](https://github.com/bpmn-io/moddle/pull/39/commits/1af0602cfb11db33cac52c66a56fa2369e90fc68) (using `eql()` for deep-equal) . The error is that`any` describes `props:any` instead of the local defined 'any' a.k.a. 'mh:any'. I only inspected `descriptor.properties` and this is ok ...
[Fixed](https://github.com/bpmn-io/moddle/pull/39/commits/36aaaccc0c6ac879ef16199786e24d21aa45a74e).

File: lib/descriptor-builder.js (hunk 6 above)
Comment (fra-pa) [reply lib/descriptor-builder.js:189]: If `p` (property to be proofed) is defined in a different namespace than `this`, then `p.name` is the local name in THAT ns and `this.propertiesByName(p.name)` might already exist (mapping to a local property). In this case, the assertion needs to hold true. This allows the property from another ns to be added (`assertNotDefined()` is called by `addNamedProperty()`).

With this patch we allow the type to reference MULTIPLE (different) properties and therefore need to distinguish `ns.name` from `ns.localName`. They can be different now.

File: lib/factory.js (hunk 7 above)
Comment (fra-pa) [reply lib/factory.js:35]: > Interesting question here: What I'd expect is to have local name and ns name attribute present.

They are present, but ...

> I.e. I can always access the value via foo["a:id"] regardless if it has a local name set or not. On the other hand I'd like to have a#id always available as the shorthand, if it is the only id property defined for an element.
>
> Likely this is only possible to accomplish via proxies, am I right about that?

For `isMany`-properties this is possible without a proxy, as we reference an object. But fundamental types (i.e. `BUILTINS`) are only accessible through `get()` when referenced with their `localName`. `a#id` on the other hand is always available through `a.id` (and `a.get('id')`of course), as long it was previously initialized (see my remarks below) - so yes, you're right.

IMO this is a compromise that we should accept. As you say, the alternative would be the introduction of proxy objects.

I hope we understand each other (please don't hesitate to keep raising questions). The test cases should hold the truth to specify desired behaviour. So we need to make sure to get them right and fully cover this topic as well as update the documentation before merging.

Btw, the current implementation also fails to access `foo['a:id']` if not initialized during creation or has a default value or was previously accessed with `get()`. I've not tested this out yet, but this seems to be the case for all properties except `foo.id`. I.e. you need to use `get()` first before `foo['a:id']` is mapped into the object.

This is somehow inconsistent and it might be a nice improvement to map property instances early (i.e. during type creation) even if they are not explicitly initialized or have a default value. I'd more like to get `undefined` instead of an unexpected exception. If you like, I'll open another issue for this.

One question in this context: should it be possible to define a default value for `isMany`-properties (e.g. `"default": [ 1, 2 ]`)? I tried this, received no exception but the value was not applied. Maybe I miss it, but there seems to be no documentation or spec for this. Should I address this in a seperate issue or am I just wrong here?

File: lib/properties.js (hunk 4 above)
Comment (fra-pa) [reply lib/properties.js]: It needs an additional parameters and I felt the function becomes more of an overhead than providing its usefulness as a wrapper with the benefit of standardizing the parameters to be passed. I removed the function in an early development stage (not without some doubt) and just kept it that way.

With your remark and a review, I [added](https://github.com/bpmn-io/moddle/pull/39/commits/e75cba86d545bc7a9a86d39c731671e12f1ba4b3) function `defineProperty()` back with an additional parameter for the property name. There are some implementation options, but I cannot find a way to get rid of this parameter.

I'm still undecided. Please take a look and decide yourself if we should keep it.

File: test/spec/properties.js (hunk 5 above)
Comment (fra-pa) [reply test/spec/properties.js]: > TODO: Add test to verify the following:
> 
>     * Package a defines `a:Foo#a`
> 
>     * Package b defines `b:Other#a`
> 
>     * Package c defines `c:MultipleInherited { a:Foo, b:Other }`
> 
>     * arrow_right What is the local descriptor for `c.MultipleInherited#a`?

I hope I got you right with [this](https://github.com/bpmn-io/moddle/pull/39/commits/d9edaf255fd398455c8ffccf9f47137b739f1437) commit, but maybe I miss your point as it seems to be redundant.

File: test/fixtures/model/properties.json [hunk 8]
```diff
@@ -122,6 +129,25 @@
           "name": "any",
           "type": "Base",
           "isMany": true
+        },
```
Comment (nikku) [inline test/fixtures/model/properties.json:132]: This does not seem to be used anymore.

File: test/fixtures/model/datatype.json [hunk 9]
```diff
@@ -7,7 +7,16 @@
       "name": "Root",
       "properties": [
         { "name": "bounds", "serialize": "xsi:type", "type": "Rect" },
-        { "name": "otherBounds", "type": "do:Rect", "serialize": "xsi:type", "isMany": true }
+        { "name": "otherBounds",
```
Comment (nikku) [inline test/fixtures/model/datatype.json:10]: Proposal: Revert to original formatting (single line) like most of the other props.

File: lib/descriptor-builder.js [hunk 10]
```diff
@@ -172,15 +181,18 @@ DescriptorBuilder.prototype.setIdProperty = function(p, validate) {
...
     throw new Error(
       'property <' + propertyName + '> already defined; ' +
-      'override of <' + definedProperty.definedBy.ns.name + '#' + definedProperty.ns.name + '> by ' +
-      '<' + p.definedBy.ns.name + '#' + p.ns.name + '> not allowed without redefines');
+        'override of <' + definedProperty.definedBy.ns.name + '#' +
```
Comment (nikku) [inline lib/descriptor-builder.js:192]: Am I blind or is this exactly equivalent to the original version, just with a different formatting?

File: test/spec/moddle.js [hunk 11]
```diff
@@ -101,7 +101,10 @@ describe('moddle', function() {
 
       expect(descriptor.ns).to.jsonEqual(expectedDescriptorNs);
       expect(descriptor.properties).to.jsonEqual(expectedDescriptorProperties);
-      expect(descriptor.propertiesByName).to.jsonEqual(expectedDescriptorPropertiesByName);
+      expect(descriptor.propertiesByName.id).to
```
Comment (nikku) [inline test/spec/moddle.js:104]: Why remove the original check here?

I'd say we could add an additional test case, if we want:

```suggestion
      expect(descriptor.propertiesByName.id).to.equal(descriptor.propertiesByName['props:id']);
```

File: test/spec/multiple-inherited-properties.js [hunk 12]
```diff
@@ -0,0 +1,383 @@
+import expect from '../expect';
+import { createModelBuilder } from '../helper';
+
+
+describe('multiple inherited properties', function() {
```
Comment (nikku) [inline test/spec/multiple-inherited-properties.js:5]: :+1: on moving the whole concern to a different file.

**Review by nikku — CHANGES_REQUESTED** [review]
Some findings inlined.

Did you already check if serialization continues to work / works with the multi-namespace support?

Cf. https://github.com/bpmn-io/moddle/pull/39#issuecomment-845336841

**Comment by fra-pa** [comment]
> Some findings inlined.

Thanks for another deep review.

> Did you already check if serialization continues to work.

I created a branch to get moddle from repository. Existing tests still green.

> works with the multi-namespace support?

Also added a first test for the writer, which failed. As with moddle, the different namespaces need to be differentiated. Had no time to dig deeper and push something of substance yet. Soon come ...

File: test/fixtures/model/properties.json (hunk 8 above)
Comment (fra-pa) [reply test/fixtures/model/properties.json:132]: Not sure, what you are referring to. Property `Root:any` or `Root:any.isMany`?

`Root:any` is used, but I explicitly added property `Root:many` with a fundamental data type to simplify testing. As it's not needed, we can remove `Root:any.isMany` for `Root:any`.

File: test/fixtures/model/datatype.json (hunk 9 above)
Comment (fra-pa) [reply test/fixtures/model/datatype.json]: I added another property and was not sure about the formatting. If you take a look at [fixtures/properties.json](https://github.com/fra-pa/moddle/blob/feat-multiple-inherited-properties/test/fixtures/model/properties.json), most of the attributes are in a seperate line, but others are defined in one line (top of the file).

It's not consistant and even though I'm not an ideologist about 80 character line limit, I apply this rule for readability and consistancy if reasonable.

[This commit](https://github.com/bpmn-io/moddle/pull/39/commits/596d002ca25e6ddb1eccc116eaad15d8beea61a3) changes the formatting back to one-liners and if you prefere this, we keep it. If you like, I'll change the formatting of all fixtures to one style.

File: lib/descriptor-builder.js (hunk 10 above)
Comment (fra-pa) [reply lib/descriptor-builder.js]: Sorry, to make the review harder than necessary with this reformatting. In the course of modifying the function, I used code, dismissed it later and forgot to also revert the formatting.

Your remark brought the function back to my mind. After taking another look, I [refactored the variables and changed the error message](https://github.com/bpmn-io/moddle/pull/39/commits/7b0a6f5098b3a1a484515a2cae8f68c2358d8b3c). I hope this makes the namespace aspect more explicit.

File: test/spec/moddle.js (hunk 11 above)
Comment (fra-pa) [reply test/spec/moddle.js:104]: > Why remove the original check here?

Because I added a property and comparing the whole structure depends on the order. That's why I created two `expect`s for this.

> I'd say we could add an additional test case, if we want:

:+1: [added](https://github.com/bpmn-io/moddle/pull/39/commits/d6f342b2bddddc89fa8f51d3a101095d0e8aaf4a)

I applied the suggestion and CI action failed with an empty error message. Fixed it [here](https://github.com/bpmn-io/moddle/pull/39/commits/66bb9e0b9e53b63ce33fd61aaa388ccc6119166d).

**Comment by nikku** [comment]
Closing in favor of https://github.com/bpmn-io/moddle/pull/45.

The moddle solution is not to support stock UML, but rather workaround the brokeness of that thing using a virtual package (https://github.com/bpmn-io/moddle/commit/401d2c1db45a0f8753989642520e24b2a8574b0f).
